algorithms/CPlusPlus/Trees/in-order-predecessor-and-successor.cpp

```diff
@@ -0,0 +1,187 @@
+#include <iostream>
+
+struct TreeNode{
+	TreeNode* left;
+	TreeNode* right;
+	int data;
+	TreeNode(const int& data): data(data), left(nullptr), right(nullptr){}
+};
+
+TreeNode* find(TreeNode* root, const int& data){
+        /**
+         * Find the node that contains the given data and
+         * return that node
+         *
+         * @params: `root` root/parent node of the tree
+         * @params: `data` data to be find in the tree
+         * @return: tree node that contains the data
+         *
+         * Average case Time Complexity: O(log(n))
+         * Worst case Time Complexity: O(n)
+         *
+         */
+
+        if(root == nullptr)         { throw std::runtime_error("Error: find() cannot find the data. The data doesn't exist."); }
+        else if(root->data == data) { return root; }
+        else if(root->data < data)  { return find(root->right, data); }
+        else                        { return find(root->left, data); }
+}
+
+void Insert(TreeNode*& root, const int& data){
+        /**
+         * Create and Insert the node in the appropriate place of the tree
+         *
+         * @params: `root` root/parent node of the tree
+         * @params: `data` data to be inserted in the tree
+         * @return: void
+         *
+         * Average case Time Complexity: O(log(n))
+         * Worst case Time Complexity: O(n)
+         *
+         */
+
+        if(root == nullptr)         { root = new TreeNode(data); }
+        else if(root->data == data) { throw std::runtime_error("The node already exist. Duplicates not allowed"); }
+        else if(root->data < data)  { Insert(root->right, data); }
+        else                        { Insert(root->left, data); }
+}
+
+int rightmost(TreeNode* node){
+	/**
+	 * Find and return the rightmost value of the sub-tree
+	 *
+         * @params: `root` root/parent node of the tree
+	 *
+	 * @return: value of the node at the rightmost end of the sub tree
+	 */
+
+	if(node->right == nullptr){ return node->data; }
+	return rightmost(node->right); 
+}
+
+int predecessor(TreeNode* node, const int& data, int value){
+	/**
+	 * Find and return the previous smallest value to data
+	 *
+         * @params: `node` root/parent node of the tree
+	 * @params: `data` data to compare with
+	 * @params: `value` holds the previous smaller value to data
+	 *
+	 * @return: previous smallest value to data
+	 */
+	
+	if(node->data < data){ return predecessor(node->right, data, node->data); }	
+	else if(node->data > data){ return predecessor(node->left, data, value); }
+	return value;
```

Comment:
Here it will return `0` if there is no predecessor.
So if it's `0` i can print there is no predecessor but what if the predecessor is actually zero what should i do then?

Reply:
instead of returning the value in the node, return a `TreeNode` type ( or a pointer to it ) which is holding/pointing to the required node in the tree and return NULL if there is no  predecessor/successor.